### PR TITLE
Determine return type of discrete pdf/logpdf based on pdf instead

### PR DESCRIPTION
**Pull request by andreasnoack**
of logpdf and test it on Rational.

This follows up on the discussion in https://github.com/JuliaStats/Distributions.jl/pull/1028#discussion_r354214245. (I like that the fixes seem to be a matter of deleting methods).

### CODE REVIEW

File: src/univariates.jl [hunk 1]
```diff
@@ -341,7 +341,6 @@ Relying on this fallback is not recommended in general, as it is prone to overfl
 """
 logpdf(d::UnivariateDistribution, x::Real) = log(pdf(d, x))
 logpdf(d::DiscreteUnivariateDistribution, x::Integer) = log(pdf(d, x))
-logpdf(d::DiscreteUnivariateDistribution, x::Real) = isinteger(x) ? logpdf(d, round(Int, x)) : partype(d)(-Inf)
```
Comment (matbesancon) [inline src/univariates.jl]: Was this method never called or was it never necessary?

File: test/categorical.jl [hunk 2]
```diff
@@ -6,6 +6,7 @@ using Test
 for p in Any[
     [0.5, 0.5],
     [0.5f0, 0.5f0],
+    [1//2, 1//2],
```
Comment (matbesancon) [inline test/categorical.jl]: The modified tests don't test the new generic zero
`pdf(d::DiscreteUnivariateDistribution, x::Real) = isinteger(x) ? pdf(d, round(Int, x)) : zero(partype(d))`

File: src/univariates.jl (hunk 1 above)
Comment (andreasnoack) [reply src/univariates.jl]: It's sufficient to have the conversion logic to `Int` in the `pdf` version which will then be called by the version in line 342.

File: test/categorical.jl (hunk 2 above)
Comment (andreasnoack) [reply test/categorical.jl]: I believe they do. The test in https://github.com/JuliaStats/Distributions.jl/blob/c89d557344591a2c046280b96b2597d019457583/test/categorical.jl#L28 will dispatch to https://github.com/JuliaStats/Distributions.jl/blob/c89d557344591a2c046280b96b2597d019457583/src/univariates.jl#L342 which will then dispatch to the modified version of https://github.com/JuliaStats/Distributions.jl/blob/c89d557344591a2c046280b96b2597d019457583/src/univariates.jl#L333. I wrote the test before the fix, so I think it's exercised.

**Pull request merged**